### PR TITLE
Implement Flake Id Generator.

### PR DESCRIPTION
**Pull request by ihsandemir**
Implemented Flake Id Generator.

### CODE REVIEW

**Comment by ihsandemir** [comment]
verify-windows

**Comment by ihsandemir** [comment]
verify-windows

**Comment by ihsandemir** [comment]
verify-windows

File: hazelcast/include/hazelcast/client/FlakeIdGenerator.h [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+             *
+             * <p>
+             * <b>Note for Flake ID Generator:</b> This method does nothing and will simply tell if the next ID
+             * will be larger than the given ID.
+             * You don't need to call this method on cluster restart - uniqueness is preserved thanks to the
+             * timestamp component of the ID.
+             *
+             * @return {@code true}, if the next ID will be larger than the supplied id
+             */
+            virtual bool init(int64_t id);
```
Comment (SaitTalhaNisanci) [inline hazelcast/include/hazelcast/client/FlakeIdGenerator.h]: Why do we have this method if it is already deprecated in Java?

File: hazelcast/include/hazelcast/client/HazelcastClient.h [hunk 2]
```diff
@@ -512,6 +512,26 @@ namespace hazelcast {
             */
             IdGenerator getIdGenerator(const std::string& name);
 
+            /**
+             * Creates a cluster-wide unique ID generator. Generated IDs are {@code long} primitive values
```
Comment (SaitTalhaNisanci) [inline hazelcast/include/hazelcast/client/HazelcastClient.h]: From this comment it seems like every time this method is called a new flake ID generator will be created, but I think if one already exists with the given name, it wont be created right? If that is the case, I think we should mention about that here.

File: hazelcast/include/hazelcast/client/FlakeIdGenerator.h (hunk 1 above)
Comment (SaitTalhaNisanci) [reply hazelcast/include/hazelcast/client/FlakeIdGenerator.h]: I guess that is because you implement `IdGenerator`

Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/FlakeIdGenerator.h]: Yes, it is there for inheritence from IdGenerator. We omitted this until the last commit where we decided that it is better to keep the inheritence in place.

File: hazelcast/include/hazelcast/client/HazelcastClient.h (hunk 2 above)
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/HazelcastClient.h]: This is the same javadoc for the Java client. if needed, Java client javadoc also needs to be changed. I changed it to: "Returns a generator that creates a cluster-wide unique IDs. "

File: hazelcast/include/hazelcast/client/config/ClientFlakeIdGeneratorConfig.h [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                 * is assigned to an object that was created much later, it will be much out of order. If you don't care
+                 * about ordering, set this value to 0.
+                 * <p>
+                 * This setting pertains only to {@link FlakeIdGenerator#newId newId} calls made on the member
+                 * that configured it.
+                 *
+                 * @param prefetchValidityMs the desired ID validity or unlimited, if configured to 0.
+                 * @return this instance for fluent API
+                 */
+                ClientFlakeIdGeneratorConfig &setPrefetchValidityMillis(int64_t prefetchValidityMillis);
```
[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/config/ClientFlakeIdGeneratorConfig.h]: Added.

File: hazelcast/include/hazelcast/client/config/ClientFlakeIdGeneratorConfig.h [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                /**
+                 * @see #setPrefetchCount(int)
+                 */
+                int32_t getPrefetchCount() const;
+
+                /**
+                 * Sets how many IDs are pre-fetched on the background when one call to
+                 * {@link FlakeIdGenerator#newId()} is made. Default is 100.
+                 *
+                 * @param prefetchCount the desired prefetch count, in the range 1..100,000.
```
[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/config/ClientFlakeIdGeneratorConfig.h]: OK, changed.

**Comment by ihsandemir** [comment]
verify-windows

**Comment by ihsandemir** [comment]
verify-windows

**Pull request merged**